### ISSUE TITLE
Allow for configuration files in .devcontainer

### ISSUE THREAD

**Issue by nerdo** [microsoft/vscode-remote-release#874]
Currently I don't seem to know a good way to have container-specific settings, tasks, and launch configurations.

The closest thing to it is to set up settings.json, tasks.json, and launch.json in the workspace `.vscode` folder, but then it applies to the workspace when you open it normally.

It would be ideal if we could have those same files in the `.devcontainer` folder and have vscode pick them up from there when the folder is opened in the container.

Currently the closest thing to it is in .devcontainer/devcontainer.json, there is a settings key which is similar to having settings.json, but i don't see anything for tasks or launch.

**Comment by Chuxel** [comment]
@nerdo There is a "settings" property in devcontainer.json that lets you put container specific settings that would normally land in settings.json.  

Tasks and launch configurations can be stored as workspace settings - and there's generally OS specific settings that can handle the differences between your local OS (mac/Windows) and Linux.

With that in mind, is there something specific you have in mind that these wouldn't cover?

**Comment by nerdo** [comment]
Yes.

In my specific use case, I’m in the process of refining a docker dev environment for our team. Setting it up with vscode using a dev container, we have some container-specific tasks (e.g. toggling PHP xdebug on and off) as well as some container specific launch settings (e.g. debugging with xdebug but also running the previous task to turn xdebug on/off automatically as pre/post tasks).

Those are just examples of some of the things I would like to include in our repository as container-specific configuration. I have set them up in my workspace in the .vscode folder, but we ignore it in git because we can’t guarantee that those things will work in everyone’s setup outside of the container.

I do understand that there is already a setup key in devcontainer.json, but it seems to me if it were broken out into its own setup.json, similar to the other workspace configuration files, that it would just homogenize configuration and the only difference (from a user standpoint) of configuring something to be container specific or workspace specific would be whether the file goes in the .vscode folder or the .devcontainer folder.

As a side note, I wasn’t able to set the shell for the container via the setup key trying to set the shell via `“terminal.integrated.shell.linux”: “/bin/bash”`. That may be a separate issue, but my thought here is that if vscode were to treat the .devcontainer directory the same way it treats .vscode for the workspace, it should just work.

**Comment by Chuxel** [comment]
@nerdo The primary goal with the `settings` property in `devcontainer.json` is to handle situations were there are properties that must be changed in a container - like absolute paths. This allows you to have a common workspace settings file and just override the few properties that are absolutely container specific.  

However, in your case, it sounds like you're looking for a container specific workspace settings in the `.vscode` folder since you are not checking the `.vscode` folder into source control.

One of the properties we provide is `postCreateCommand` which is intended to allow you to execute setup steps after the dev container has been created.  One way to do what you are talking about would be to have a folder with the container specific `task.json`, `launch.json`, `settings.json`, etc in it in the source tree somewhere and use the postCreateCommand to move it to the correct location.

Say you've put these in a folder called `setup/vscode-config` and you have `.vscode` in your `.gitignore` file.  Your post create command could just look like this:

```json
"postCreateCommand": "cp setup/vscode-config .vscode"
```
This is run after the container is created and your source folder has been mounted but before VS Code attaches to it, so it would pick up the contents of the .vscode folder at that point.

Would that work in your situation?

**Comment by nerdo** [comment]
I understand the reason for the `settings` propety. I am using settings in this situation for that reason to a certain degree, but also trying to set up some nice-to-haves that are container-specific.

It sounds like something like that might work, but I wouldn't want to overwrite what is in `.vscode`. Even though it's in our `.gitignore`, the assumption is that everyone will have their own, personal settings there.

I do, however, recall seeing settings being stored inside a running container in `/root/.vscode-server/data/Machine/`. Setting it up to copy there might work... that is, assuming that the `postCreateCommand` runs from _within_ the container.

Is that the case?

**Comment by Chuxel** [comment]
@nerdo Yep, `postCreateCommand` does run within the container and it happens after the source code is mounted, so you can even use it to run shell scripts in your source tree in the container (e.g. `bash scripts/setup-some-stuff.sh`).

Unfortunately, while the path you listed is where the settings file is located (or under `.vscode-server-insiders` if you use insiders), that won't help with your `tasks.json` and `launch.json`. This location is the similar to the user location of `~/.vscode/User/settings.json` or `%APPDATA%/Code/User/settings.json` locally. The `tasks.json` and `launch.json` files only work in your workspace folder only. (You can, however, use variables to manage things like paths, and there's OS specific properties in many launch configs that you can add to handle Windows/Mac/Linux differences in the same file of course.)

Is there a particular example you have in mind?

**Comment by nerdo** [comment]
The particular example I have in mind is this. I'm on a dev team and have been working on a dockerized dev setup, which work well with vscode and its new remote capabilities.

Everyone on the team has their own vscode settings though, including but not limited to tasks and other preferences, but for now I really just want to focus on tasks and launch settings.

Since it's a dockerized container, it is very easy to set up tasks that are ready-to-go without any configuration or fiddling. I'm looking for a way to provide tasks and launch settings (like starting dev servers, running build scripts, launching debuggers) without overwriting other custom tasks that a developer may have set up.
